python/tvm/relax/transform/legalize_ops/index.py

```diff
@@ -59,3 +60,39 @@ def _strided_slice(bb: BlockBuilder, call: Call) -> Expr:
         call.attrs.axes,
         slice_mode="end",
     )
+
+
+@register_legalize("relax.dynamic_strided_slice")
+def _dynamic_strided_slice(bb: BlockBuilder, call: Call) -> Expr:
+    # 1. Insert shape function
+    output_shape = bb.normalize(
+        bb.call_te(
+            topi.shape_func_dynamic_strided_slice,
+            call.args[0],
+            call.args[1],
+            call.args[2],
+            call.args[3],
```

Comment:
Two questions:

* Why `shape_func_dynamic_strided_slice` needs to be implemented in topi / c++? This seems like a one-off function, why not just create `te.compute` here in python?
* Are we introducing a standard API for shape func (like Relay), or is each op going to pick any API as it sees fit (like this example)? In the latter case I don't see any reason shape funcs need to belong to topi.

Reply:
> Are we introducing a standard API for shape func (like Relay), or is each op going to pick any API as it sees fit (like this example)? In the latter case I don't see any reason shape funcs need to belong to topi.

The latter case. 

Great point. I put it in TOPI-side because I viewed TOPI as the registry of TEs and developers can easily check related implementations in the single file. I don't have any strong preference here, so I'm open to hear what folks prefer.

Reply:
Since there is no standard API for shape func, and there is no "registration" for shape funcs specifically, I think it is better not to introduce a rigid convention, in particular shape funcs be implemented in topi / c++. 

Since they are only used by the legalizer, I view shape funcs in Relax simply as implementation details, one-off function. So developers should be able to express data-dependent output shape computation logic in any way in anywhere, including TVMScript.

Reply:
I think this is very good point. we can continue this discussion here: https://github.com/apache/tvm/pull/14548#discussion_r1162233085